addons/survey/controllers/survey_session_manage.py

```diff
@@ -176,28 +174,21 @@ def survey_session_code(self, **post):
     @http.route('/s/<string:session_code>', type='http', auth='public', website=True)
     def survey_start_short(self, session_code):
         """" Redirects to 'survey_start' route using a shortened link & token.
-        We match the session_code for open surveys.
+        Shows an error message if the survey is not valid.
         This route is used in survey sessions where we need short links for people to type. """
+        survey_info = self._fetch_from_session_code(session_code)
 
-        survey = self._fetch_from_session_code(session_code)
-        if survey and survey.session_state in ['ready', 'in_progress']:
-            return request.redirect("/survey/start/%s" % survey.access_token)
-
-        return request.redirect("/s")
+        if 'error' in survey_info:
+            survey_info['session_code'] = session_code
+            return request.render('survey.survey_session_code', survey_info)
+        return request.redirect(survey_info['survey_url'])
 
     @http.route('/survey/check_session_code/<string:session_code>', type='json', auth='public', website=True)
     def survey_check_session_code(self, session_code):
         """ Checks if the given code is matching a survey session_code.
         If yes, redirect to /s/code route.
-        If not, return error. The user is invited to type again the code. """
-        survey = self._fetch_from_session_code(session_code)
-        if survey:
-            if survey.session_state in ['ready', 'in_progress']:
-                return {"survey_url": "/survey/start/%s" % survey.access_token}
-            else:
-                return {"error": "survey_session_closed"}
-
-        return {"error": "survey_wrong"}
+        If not, return error. The user is invited to type again the code."""
+        return self._fetch_from_session_code(session_code)
```

Comment:
If we change the fetch from session code returned result as proposed, would become
```
_survey, survey_info = self._fetch_from_session_code(session_code)
return survey_info
```